File: Example/PHP/example/Flarum/Plugins/NextSpanPlugin.php
<?php
#-------------------------------------------------------------------------------
# Copyright 2019 NAVER Corp
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy
# of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.
#-------------------------------------------------------------------------------

namespace Plugins;
use Plugins\Candy;

class NextSpanPlugin extends Candy
{

///@hook:Laminas\Diactoros\Response\RedirectResponse::__construct
    public function onBefore(){
/*
        $headers = &$this->args[2];
        if(PerRequestPlugins::instance()->traceLimit()){
            $headers['Pinpoint-Sampled'] = 's0';
            return ;
        }
        $headers['Pinpoint-Sampled'] ='s1';
        $headers['Pinpoint-Flags'] ='0';
        $headers['Pinpoint-Papptype'] ='1500';
        $headers['Pinpoint-Pappname'] =PerRequestPlugins::instance()->app_name;

        $headers['Pinpoint-Host'] = $this->args[0];
        $headers['Pinpoint-Traceid'] =PerRequestPlugins::instance()->tid;
        $headers['Pinpoint-Pspanid'] =PerRequestPlugins::instance()->sid;
        $nsid = PerRequestPlugins::instance()->generateSpanID();
        $headers['Pinpoint-Spanid'] =$nsid;
*/
        pinpoint_add_clue("dst",$this->getHostFromURL($this->args[0]));
        pinpoint_add_clue("stp",PINPOINT_PHP_REMOTE);
        pinpoint_add_clue('nsid',PerRequestPlugins::instance()->getCurNextSpanId());
        pinpoint_add_clues(HTTP_URL,parse_url($this->args[0],PHP_URL_PATH));
    }

    public function onEnd(&$ret){
        pinpoint_add_clues(HTTP_STATUS_CODE, string($this->who->getStatusCode()));
    }

    public function onException($e){
        pinpoint_add_clue("EXP",$e->getMessage());
    }

    function getHostFromURL(string $url)
    {
        $urlAr   = parse_url($url);
        $retUrl = '';
        if(isset($urlAr['host']))
        {
            $retUrl.=$urlAr['host'];
        }

        if(isset($urlAr['path'])){
            $retUrl.=$urlAr['path'];
        }

        if(isset($urlAr['port']))
        {
            $retUrl .= ":".$urlAr['port'];
        }

        return $retUrl;
    }
}

File: plugins/PY/pinpointPy/libs/_requests/NextSpanPlugin.py
# ------------------------------------------------------------------------------
#  Copyright  2020. NAVER Corp.                                                -
#                                                                              -
#  Licensed under the Apache License, Version 2.0 (the "License");             -
#  you may not use this file except in compliance with the License.            -
#  You may obtain a copy of the License at                                     -
#                                                                              -
#   http://www.apache.org/licenses/LICENSE-2.0                                 -
#                                                                              -
#  Unless required by applicable law or agreed to in writing, software         -
#  distributed under the License is distributed on an "AS IS" BASIS,           -
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.    -
#  See the License for the specific language governing permissions and         -
#  limitations under the License.                                              -
# ------------------------------------------------------------------------------


from pinpointPy import Common, Helper, Defines, pinpoint
from urllib.parse import urlparse


class NextSpanPlugin(Common.PinTrace):

    def __init__(self, name):
        super().__init__(name)

    @staticmethod
    def isSample(*args, **kwargs):
        sampled, parentId, _, _ = Common.PinTrace.isSample(args, kwargs)
        headers = {}
        if 'headers' in kwargs:
            headers = kwargs['headers']
        if Defines.PP_HEADER_PINPOINT_SAMPLED in headers:
            return False, parentId, args, kwargs
        # pull out headers
        if sampled:
            url = args[1]
            target = urlparse(url).netloc
            if pinpoint.get_context(Defines.PP_HEADER_PINPOINT_SAMPLED, parentId) == "s1":
                Helper.generatePinpointHeader(
                    target, headers, traceId=parentId)
        else:
            headers[Defines.PP_HEADER_PINPOINT_SAMPLED] = Defines.PP_NOT_SAMPLED
        kwargs['headers'] = headers
        return sampled, parentId, args, kwargs

    def onBefore(self, parentId, *args, **kwargs):
        traceId, _args, _kwargs = super().onBefore(parentId, *args, **kwargs)
        url = _args[1]
        target = urlparse(url).netloc
        ###############################################################
        pinpoint.add_trace_header(
            Defines.PP_INTERCEPTOR_NAME, self.getUniqueName(), traceId)
        pinpoint.add_trace_header(
            Defines.PP_SERVER_TYPE, Defines.PP_REMOTE_METHOD, traceId)
        pinpoint.add_trace_header_v2(Defines.PP_ARGS, url, traceId)
        pinpoint.add_trace_header_v2(Defines.PP_HTTP_URL, url, traceId)
        pinpoint.add_trace_header(Defines.PP_DESTINATION, target, traceId)
        ###############################################################
        return traceId, _args, _kwargs

    def onEnd(self, traceId, ret):
        ###############################################################
        pinpoint.add_trace_header(Defines.PP_NEXT_SPAN_ID, pinpoint.get_context(
            Defines.PP_NEXT_SPAN_ID, traceId), traceId)
        if ret:
            pinpoint.add_trace_header_v2(
                Defines.PP_HTTP_STATUS_CODE, str(ret.status_code), traceId)
            pinpoint.add_trace_header_v2(Defines.PP_RETURN, str(ret), traceId)
        ###############################################################
        super().onEnd(traceId, ret)
        return ret

    def onException(self, traceId, e):
        pinpoint.add_trace_header(Defines.PP_ADD_EXCEPTION, str(e), traceId)
